### DISCUSSION TITLE
How can i debug casadi-c++ code in VSCode?

### DISCUSSION THREAD

**Discussion by KirollosMedhat** [casadi/casadi#3128] [Q&A]
I am trying to debug my casadi-c++ code but the debugger just fails to stop at breakpoints, fails to track local variables, and also is failing to watch variables and objects.

Did anyone try to debug casadi-c++ code before?

**Comment by jaeandersson** [comment]
I can use the VS Code debugger. In Windows, it works for both visual studio compiler and with mingw. Did you compile CasADi from sources?

**Reply by KirollosMedhat** [reply]
im on ubuntu and i figured out the problem in the cmakelists.txt file there was that line:
set(CMAKE_CXX_FLAGS "-Wall -Wextra -pthread")
i added the -g flag it became: 
set(CMAKE_CXX_FLAGS "-Wall -Wextra -pthread -g")
and now the debugging session is working fine.. thanks.

**Reply by KirollosMedhat** [reply]
but may i ask you, i am on ubuntu because i only found casadi installation tutorial on ubuntu, if there's any on windows can u send it to me?

**Reply by jaeandersson** [reply]
I don't think you should manually add "-g". Are you building a "Release" build instead of a "Debug" build? "-g" should be added automatically for Debug. From command-line, something like -DCMAKE_BUILD_TYPE=Debug.

The source build instructions are woefully out-of-date. There is an open issue to fix them here: #3061. I'm sketching on an easier way to build from sources, using CMake presets. The latest release has a first version of a CMakePresets.json, so in principle you could try to build a minimal development version using:
```
cmake .. --preset=debug
```
Or a more typical Linux build using:
```
cmake .. --preset=devel
```
For Windows there are the presets "windevel" and "windevel_cl" for MinGW and Visual C++ respectively.

Building just the CasADi core, the Python front end and some common interfaces (like IPOPT), is usually quite easy. It can get very complicated if you want to build the MATLAB/Octave front-end and a lot of third-party interfaces.

**Reply by jaeandersson** [reply]
If you are able to test and give feedback on some new source build instructions, let me know. It would make it easier to complete #3061

**Reply by Dave-Poissant** [reply]
This question is still relevant but in another way in my opinion. It would be usefulll to have a way to debug expressions during execution like it is on Matlab. I think I've understood that CasADI builds graphical trees for symbolic expression, which doesn't seem accessible. It could be usefull to send something like a casadi::Function to the solver definition that gets called when it exists during optimization (every iteration) where we can access solver expressions like cost, state matrices, constraints, etc.

Am I missing something in the C++ API that already makes this possible? I'm new to this Library so I might be.

**Reply by brunoml5** [reply]
Hi, @Dave-Poissant! Have you tried using the `monitor` option from the solver? Maybe this is what you're looking for

**Reply by brunoml5** [reply]
Or maybe you can use the callback feature from the opti interface ([link to docs](https://web.casadi.org/docs/#extras))
